shortfin/tests/apps/llm/cpu_llm_server_test.py

```diff
@@ -0,0 +1,156 @@
+import pytest
+import subprocess
+import time
+import requests
+import os
+import json
+import uuid
+import tempfile
+import shutil
+
+from transformers import AutoTokenizer
+
+BATCH_SIZES = [1, 4]
+cpu_settings = {
+    "device_flags": ["-iree-hal-target-backends=llvm-cpu"],
+    "device": "local-task",
+}
+gpu_settings = {
+    "device_flags": ["-iree-hal-target-backends=rocm", "--iree-hip-target=gfx1100"],
+    "device": "hip",
+}
+settings = cpu_settings
+
+
+@pytest.fixture(scope="module")
+def model_test_dir(tmp_path_factory):
+    tmp_dir = tmp_path_factory.mktemp("cpu_llm_server_test")
+    try:
+        # Download model if it doesn't exist
+        model_path = os.path.join(tmp_dir, "open-llama-3b-v2-f16.gguf")
+        if not os.path.exists(model_path):
+            subprocess.run(
+                f"huggingface-cli download --local-dir {tmp_dir} SlyEcho/open_llama_3b_v2_gguf open-llama-3b-v2-f16.gguf",
+                shell=True,
+                check=True,
+            )
+
+        # Set up tokenizer if it doesn't exist
+        tokenizer_path = os.path.join(tmp_dir, "tokenizer.json")
+        if not os.path.exists(tokenizer_path):
+            tokenizer = AutoTokenizer.from_pretrained(
+                "openlm-research/open_llama_3b_v2"
+            )
+            tokenizer.save_pretrained(tmp_dir)
+
+        # Export model if it doesn't exist
+        mlir_path = os.path.join(tmp_dir, "model.mlir")
+        config_path = os.path.join(tmp_dir, "config.json")
+        if not os.path.exists(mlir_path) or not os.path.exists(config_path):
+            bs_string = ",".join(map(str, BATCH_SIZES))
+            subprocess.run(
+                [
+                    "python",
+                    "-m",
+                    "sharktank.examples.export_paged_llm_v1",
+                    f"--gguf-file={model_path}",
+                    f"--output-mlir={mlir_path}",
+                    f"--output-config={config_path}",
+                    f"--bs={bs_string}",
+                ],
+                check=True,
+            )
+        # Compile model if it doesn't exist
+        vmfb_path = os.path.join(tmp_dir, "model.vmfb")
+        if not os.path.exists(vmfb_path):
```

Comment:
Caching in tests can be tricky. If the user updates the compiler, they might want the test to use the newer version of the tool.

Please be very deliberate here.
* Which files are going to persistent paths and which are going to temporary paths?
* Should the developer running this test be able to control the cache behavior, e.g. by explicitly clearing cached files at the start or end of the test?
* Log when using a cached version
* Log when compiling (especially if compilation takes more than 10 seconds)
* Log the full command being run

Reply:
That's a good point. Looking at it again, I think it'd be "safe" to cache `open-llama-3b-v2-f16.gguf` and `tokenizer.json`.

Unless I'm wrong, I don't think those files will change, and take up the bulk of the time when downloading collaterals to run the test.

That leaves the files produced from `sharktank.examples.export_paged_llm_v1` and `iree-compile` uncached, which seems like it could add some value as an e2e test and catch any potential errors in those two necessary commands.

Reply:
Added cachine for model parameter and tokenizer files, based on HF_HOME dir